### DISCUSSION TITLE
Handling large network and error message on HPC

### DISCUSSION THREAD

**Discussion by osupplisson** [davidbolin/MetricGraph#13] [Q&A] [answered]
Hi,

First of all, thank you very much for your wonderful package and complete documentation.

I am trying to set up an LGCP model to study disease incidence in a South-East Asian city.

The sf spatial objects that I am trying to turn into a metric graph is built from a subset of roads from the city, as providedby Open Street Maps. 

There are 233,866 linestrings, the minimum length is 0.001870287 meters and the maximum length is 1028.852 meters.

The SF object weights 63.4 MB in R.

I have been facing the following issues:

- On my laptop (60GO RAM, 13th Gen Intel(R) Core(TM) i7, Window 11), R4.4.1.: using `metric_graph$new()`  crashes the laptop due to lack of RAM. When cropping the SF to a smaller area, it works.

Q1. Is there any strategy to reduce the memory cost of creating a graph? Maybe something like building the network for subspaces and then combining everything in the end?

- I gave it a try with two different HPC (320GO RAM and 1,900GO RAM, R4.4.0). I got the following error (with the full sf or when cropping it):

> Error in initialize(...) : 
> If one of 'vertex_unit' or 'length_unit' is NULL, and the edges are not sf nor sp objects, then the other must also be NULL.

Q2. I am a bit puzzled because all geometries are non-empty and the same code works on my laptop but not on the HPCs, do you have any idea about what is going wrong?

For both the laptops and HPCs I am using the version 1.3.0.9000.

Q3. I take the opportunity of this post to ask a third question: the vignette about LGCP model has only one example with R-INLA, is there any support for this kind of model with inlabru? Especially, I am stuck with what should be passed to the `domain=list(...)` in the `like()` function. Following the classical inlabru syntax, it should something like `domain=list(loc = graph$mesh)`, but all my attempt have failed so far.

Best,
Olivier

**Comment by AlexandreBSimas** [comment, marked as answer]
Hi Olivier,

We are very happy that you are enjoying the package!

For handling very large metric graphs our suggestion is to construct the metric graph using the following arguments `perform_merges=FALSE` and ` merge_close_vertices = FALSE`.

Furthermore, after you creating the metric graph, we also suggest you to prune the vertices of degree 2, with the `prune_vertices` method, like `graph$prune_vertices()`.

These should help a lot in your case. 

Another thing that you need to be careful is that version 1.3.0.9000 can be a bit misleading, because it simply means that it is *some* development version after 1.3.0. So might be that you have an older version. So, I would suggest you to reinstall it, say with, 
`devtools::install_github("davidbolin/MetricGraph", ref="devel")`,  to make sure you really have the most recent version.

Ahh, with respect to inlabru, we will add shortly. I can let you know here when we finish adding it. We will also add soon some (non-separable) space-time models.

Please, also install the newest version of `rSPDE` as we also have improved the rspde model creation (in speed and memory).

So, besides doing that for the HPC, you should also try in your laptop if you want, as it should be much lighter. 

Please, let me know if you still encounter issues after trying these things (reinstall the metric graph package in the hpc to make sure you indeed have the most recent version, build the metric graph object without the merges, then performing prunes, then proceeding with the rest). 

Best,
Alexandre

**Reply by osupplisson** [reply]
Hi Alexandre,

Thank you very much for your help, it is indeed working easily with the last releases of the packages and the suggested arguments (both on the HPC and my laptop) ! 

Regarding the support for LGCP with inlabru, I will have a regular look at the package homepage/commits but just in case I miss it that would indeed be awesome if you could keep me posted on this thread. I have never tried SPDE models with INLA, I'll give it a try while waiting for the update.

Again, thank you for your help and your work.

Best,
Olivier
